### ISSUE TITLE
Is EndAlignment not a Layout cop?

### ISSUE THREAD

**Issue by bquorning** [bbatsov/rubocop#4704]
In #4278, a bunch of whitespace cops were moved into the new “Layout department”. Since, I have wondered why `Lint/EndAlignment` wasn’t moved along with them.

**Comment by jonas054** [comment]
`Layout` was added mostly to off-load `Style`, and it didn't occur to me that there were cops dealing with whitespace in other departments too. But it would make sense to move it. Especially since it makes `rubocop --auto-correct --only Layout` work better. We can still configure it to be on `Warning` level.

**Comment by bquorning** [comment]
I think the Layout namespace makes a lot of sense. My .rubocop.yml files almost entirely consist of changing the enforced styles of Layout cops.

I didn’t really know what was the difference between Lint and Style, and I have never used the severity level feature of RuboCop. I just now read this part of the manual:

> Each cop has a default severity level based on which department it belongs to. The level is `warning` for `Lint` and `convention` for all the others. Cops can customize their severity level. Allowed params are `refactor`, `convention`, `warning`, `error` and `fatal`.

I actually think `warning` is the wrong level for `Lint/EndAlignment`, and would propose to just move the cop into `Layout` and be on Convention level.

Probably `Lint/BlockAlignment`, `Lint/ConditionPosition`, and `Lint/DefEndAlignment` can be moved, too?

**Comment by jonas054** [comment]
These cops are in the `Lint` department to indicate that the offenses are possibly bugs, and not just bad style choices. For that reason I want the level to be `warning`, but I support the idea of moving all of them.

**Comment by bquorning** [comment]
I don’t immediately see how any of the offenses caught by these 4 cops might be a bug. Can you give any examples?

**Comment by Drenmi** [comment]
I was looking through the style guide (and the internet) but can't find any examples of bugs caused by misaligned `end` without finding anything. Could be that it's from older Ruby versions, or that it was just put there arbitrarily?

I agree it would be convenient to have these together in `Layout`.

**Comment by mikegee** [comment]
Perhaps it is a warning in Rubocop because it is a warning in Ruby:

```
% cat test/test.rb
begin
  foobar
  end
```
```
% ruby -cW2 test/test.rb
test/test.rb:3: warning: mismatched indentations at 'end' with 'begin' at 1
Syntax OK
```

**Comment by bbatsov** [comment]
> I was looking through the style guide (and the internet) but can't find any examples of bugs caused by misaligned end without finding anything. Could be that it's from older Ruby versions, or that it was just put there arbitrarily?

MRI treats this as a warning and that's the reason why it's a lint cop in RuboCop as well. I think the reasoning was that you might not perceive properly which scope are you operating in because of the broken alignment.

**Comment by bbatsov** [comment]
> Perhaps it is a warning in Rubocop because it is a warning in Ruby:

Exactly.

**Comment by mikegee** [comment]
> Exactly.

I'm so sorry for the confusion, but the code I posted is not an offense for `EndAlignment`. [This "bad" example](https://github.com/bbatsov/rubocop/blob/master/lib/rubocop/cop/lint/end_alignment.rb#L22-L25) from the `Lint/EndAlignment` docs raises no warnings from Ruby:

```ruby
variable = if true
    end
```

I got mixed up by the names of these cops.

**Comment by bquorning** [comment]
> I have wondered why `Lint/EndAlignment` wasn’t moved [to `Layout` namespace]

I believe the same question applies to `Lint/BlockAlignment`.

**Comment by bquorning** [comment]
I see Ruby warnings for `Lint/EndAlignment` and `Lint/DefEndAlignment` offenses, but not for `Lint/BlockAlignment` and `Lint/ConditionPosition`:

`» cat test.rb`
```rb
# Lint/EndAlignment
class Test
  end

# Lint/DefEndAlignment
def test
  end

# Lint/BlockAlignment
test do
  end

# Lint/ConditionPosition
if
true
"true"
elsif
false
"false"
end
```

```
» ruby -w test.rb
test.rb:3: warning: mismatched indentations at 'end' with 'class' at 2
test.rb:7: warning: mismatched indentations at 'end' with 'def' at 6
```

**Comment by bquorning** [comment]
Also relevant comment: https://github.com/bbatsov/rubocop/pull/1789#issuecomment-92308357

> The end alignment cops are in the `Lint` category because the bad indentation could be something more serious than just a style issue. It could be a mistake in which keyword you think you're matching with the `end`. (`ruby -w` warns for these too, by the way.) So for this reason I don't think we can add auto-correct for these cops.
